### ISSUE TITLE
fix(security): unsafe tarfile.extractall, no checksums, plain HTTP in datasets/downloader.py

### ISSUE THREAD

**Issue by mvillmow** [HomericIntelligence/Hephaestus#478]
## Problem

Three related security defects in `hephaestus/datasets/downloader.py`:

1. **Unsafe tarfile extraction (CWE-22 path traversal).** `tf.extractall(output_path)`
   at lines 316 (CIFAR-10) and 437 (CIFAR-100) extracts archives without a path
   filter. A tarball whose members contain `../` or absolute paths can write outside
   the target directory. Python 3.12 deprecated unfiltered `extractall` for this
   reason.

2. **No checksum verification.** Downloaded tarballs and gzips are extracted
   unconditionally. A MITM or compromised CDN can substitute malicious content
   undetected.

3. **Plain HTTP for Fashion-MNIST** (`downloader.py:225`):
   `http://fashion-mnist.s3-website.eu-central-1.amazonaws.com` — no TLS, content is
   trivially substitutable. SECURITY.md claims "All dataset downloads use HTTPS"
   (false today).

## Fix

1. Pass `filter='data'` to both `tf.extractall(...)` calls (Python 3.12+).
2. Add a per-file MD5 verification step (matching torchvision's canonical values),
   re-downloading on mismatch and failing on a second mismatch.
3. Change the Fashion-MNIST base URL to `https://`.

## Acceptance

- No unfiltered `tarfile.extractall` calls in the codebase.
- Downloaded archives have their MD5 verified against a known-good per-file map;
  mismatches fail loudly with the file deleted.
- All dataset downloads use HTTPS.

Found by the strict 1.0-readiness audit (Section 8, MAJOR — security).

**Comment by mvillmow** [comment]
# Implementation Plan


> [!WARNING]
> **NOGO-EXHAUSTED** — The strict review loop ran all 3 iterations without an unambiguous GO verdict. This plan was posted for operator review but **should not be implemented** until a human approves it.

Reviewer's MAJOR finding also confirmed: `zf.extract(name, output_path)` at line 571 — `zipfile.extract()` does sanitize traversal in modern Python 3.10+ (since CVE-2007-4559 wasn't fully addressed until 3.6.2; current `zipfile.extract` does normalize and strip leading slashes per the Python source), but member names containing `..` segments after the leading slash strip have historically been a gray area, and the issue's "no unfiltered extraction" principle plus `_KNOWN_MD5S` integrity gap both apply to EMNIST regardless. The MD5 integrity gap for `gzip.zip` is the clear violation.

# Implementation Plan: Issue #478 — Follow-up Remediation (R2)

**Correction**: My R1 plan dismissed the R0 review as a timeout based on the text in the prompt. The reviewer is correct that this was wrong — the R0 findings describe real, verifiable defects in the merged code. I confirmed both findings by reading the file directly:

- **CRITICAL**: `_DATASET_MD5` (downloader.py:54–57) maps the four shared MNIST/Fashion-MNIST filenames to Fashion-MNIST hashes only. `MNISTDownloader` (line 222) and `FashionMNISTDownloader` (line 280) both download files named `train-images-idx3-ubyte.gz` etc. `_verify_or_remove(path, filename)` keys by basename only → MNIST downloads always fail verification, the file is unlinked, retries exhaust, `download_with_retry` returns `False`. MNIST is broken.
- **MAJOR**: EMNIST `gzip.zip` (downloader.py:548–571) has no MD5 entry in `_DATASET_MD5` and is extracted with `zf.extract(name, output_path)` with no integrity gate. The acceptance criterion "Downloaded archives have their MD5 verified against a known-good per-file map" is unmet for EMNIST.

Issue #478 must be reopened, or a follow-up issue filed and a remediation PR landed.

## 1. Objective

Close the two defects introduced/missed by the original #478 fix:

1. Fix MNIST verification so MNIST hashes are checked against MNIST files and Fashion-MNIST hashes against Fashion-MNIST files, despite shared filenames.
2. Extend MD5 verification + safe extraction to EMNIST (`gzip.zip` archive and its inner members).

## 2. Approach

**Key design decision: key the MD5 map by (dataset, filename), not filename alone.** The shared-filename collision is intrinsic to MNIST and Fashion-MNIST using upstream-identical names; any single-key map will collide. Two viable shapes:

- **Option A (chosen)**: Restructure `_DATASET_MD5: dict[str, dict[str, str]]` keyed by dataset id (`"mnist"`, `"fashion_mnist"`, `"cifar10"`, `"cifar100"`, `"emnist"`), inner map filename → md5. Each downloader passes its own dataset id when calling verification.
- Option B: Per-downloader class constant + module-level helper that takes the map as a parameter. Marginally less central but achieves the same disambiguation.

Option A keeps a single source of truth and one helper signature change; the inner-dict shape mirrors torchvision's organization-by-dataset.

**Helper signature change**: `_verify_or_remove(path: Path, dataset: str, filename: str) -> bool` and `download_with_retry(..., dataset: str | None = None)` so callers can opt in. Backward-compatible: when `dataset is None`, no verification is attempted (matching today's "no checksum recorded — skipping" path) and a warning is logged.

**EMNIST integrity**: Add the published `gzip.zip` MD5 under `"emnist"`. `EMNISTDownloader.download_emnist` passes `dataset="emnist"` to `download_with_retry`. For the inner zip extraction, switch from `zf.extract(name, output_path)` to a validated extract that resolves each member path under `output_path` and rejects any member whose resolved path escapes the target — the canonical safe-extraction pattern for `zipfile` (mirroring what `tarfile filter="data"` does for tar). A small helper `_safe_zip_extract(zf, member, dest)` keeps the logic in one place.

**MNIST MD5 source**: Yann LeCun's site publishes file sizes but the canonical MD5s are catalogued in torchvision's `MNIST.resources` (and PyTorch issue threads). Use those values — they are the de-facto canonical reference, same source as the Fashion-MNIST hashes already in the file.

## 3. Files to Create

None. Tests appended to the existing `tests/unit/datasets/test_downloader.py`.

## 4. Files to Modify

### `hephaestus/datasets/downloader.py`

**Restructure `_DATASET_MD5` (lines 49–58)** to:
```python
_DATASET_MD5: dict[str, dict[str, str]] = {
    "cifar10": {"cifar-10-python.tar.gz": "c58f30108f718f92721af3b95e74349a"},
    "cifar100": {"cifar-100-python.tar.gz": "eb9058c3a382ffc7106e4002c42a8d85"},
    "mnist": {
        "train-images-idx3-ubyte.gz": "f68b3c2dcbeaaa9fbdd348bbdeb94873",
        "train-labels-idx1-ubyte.gz": "d53e105ee54ea40749a09fcbcd1e9432",
        "t10k-images-idx3-ubyte.gz":  "9fb629c4189551a2d022fa330f9573f3",
        "t10k-labels-idx1-ubyte.gz":  "ec29112dd5afa0611ce80d1b7f02629c",
    },
    "fashion_mnist": {
        "train-images-idx3-ubyte.gz": "8d4fb7e6c68d591d4c3dfef9ec88bf0d",
        "train-labels-idx1-ubyte.gz": "25c81989df183df01b3e8a0aad5dffbe",
        "t10k-images-idx3-ubyte.gz":  "bef4ecab320f06d8554ea6380940ec79",
        "t10k-labels-idx1-ubyte.gz":  "bb300cfdad3c16e7a12a480ee83cd310",
    },
    "emnist": {"gzip.zip": "58c8d27c78d21e728a6bc7b3cc06412e"},
}
```
(Implementer must confirm MNIST and EMNIST hashes against torchvision's `MNIST.resources` / `EMNIST.url`/`md5` before committing.)

**Update `_verify_or_remove` (lines 70–93)** signature to `(path: Path, dataset: str, filename: str) -> bool`; look up `_DATASET_MD5.get(dataset, {}).get(filename)`.

**Update `download_with_retry`** (the call site that invokes `_verify_or_remove`) to accept `dataset: str | None = None` and forward it. When `None`, skip verification (preserves the "no checksum recorded" path).

**Update each downloader to pass its dataset id:**
- `MNISTDownloader.download_mnist` → `self.download_with_retry(gz_filename, gz_path, dataset="mnist")`
- `FashionMNISTDownloader.download_fashion_mnist` → `dataset="fashion_mnist"`
- `CIFAR10Downloader.download_cifar10` → `dataset="cifar10"`
- `CIFAR100Downloader.download_cifar100` → `dataset="cifar100"`
- `EMNISTDownloader.download_emnist` (lines 552–561) → `dataset="emnist"`

**Add `_safe_zip_extract` helper:**
```python
def _safe_zip_extract(zf: "zipfile.ZipFile", member: str, dest: Path) -> None:
    target = (dest / member).resolve()
    if not str(target).startswith(str(dest.resolve()) + os.sep) and target != dest.resolve():
        raise ValueError(f"Refusing zip member outside target: {member!r}")
    zf.extract(member, dest)
```
Replace `zf.extract(name, output_path)` at line 571 with `_safe_zip_extract(zf, name, output_path)`.

### `tests/unit/datasets/test_downloader.py`

Add regression tests:

1. `test_dataset_md5_mnist_and_fashion_mnist_have_distinct_hashes` — for each shared filename, assert `_DATASET_MD5["mnist"][f] != _DATASET_MD5["fashion_mnist"][f]`. (This test would have caught the original collision.)
2. `test_dataset_md5_covers_all_downloaders` — assert keys `{"mnist","fashion_mnist","cifar10","cifar100","emnist"}` all present and non-empty.
3. `test_verify_or_remove_uses_dataset_keyed_lookup` — write a file with MNIST content, call `_verify_or_remove(path, "mnist", "train-images-idx3-ubyte.gz")` with monkeypatched map → True; call with `dataset="fashion_mnist"` and Fashion-MNIST hash → False (and file unlinked).
4. `test_mnist_downloader_passes_dataset_id` — mock `download_with_retry`, invoke `download_mnist`, assert each call received `dataset="mnist"`.
5. `test_fashion_mnist_downloader_passes_dataset_id` — symmetric.
6. `test_emnist_downloader_passes_dataset_id_and_uses_safe_extract` — mock `zipfile.ZipFile`, invoke `download_emnist`, assert `_safe_zip_extract` was used (not raw `zf.extract`) and `download_with_retry` got `dataset="emnist"`.
7. `test_safe_zip_extract_rejects_traversal` — construct a `zipfile.ZipFile` in a tmp dir containing a member named `"../escape.txt"`, assert `_safe_zip_extract` raises `ValueError` and no file is written outside `dest`.

## 5. Implementation Order

1. Restructure `_DATASET_MD5` to the nested shape; copy values; add MNIST + EMNIST hashes (verify against torchvision source).
2. Change `_verify_or_remove` signature; add `dataset` parameter; update lookup.
3. Thread `dataset` through `download_with_retry`.
4. Update all five downloader classes to pass their dataset id.
5. Add `_safe_zip_extract`; replace `zf.extract` call at line 571.
6. Write the 7 regression tests.
7. `pixi run pytest tests/unit/datasets/test_downloader.py -v` — all green.
8. `pixi run ruff check hephaestus/ tests/` and `pixi run mypy hephaestus/datasets/`.
9. Manual sanity: run `python -m hephaestus.datasets.downloader mnist /tmp/mnist_test` end-to-end against the real upstream — must succeed (validates the MNIST hashes are correct).
10. Signed commit, push branch (e.g., `478-fix-mnist-collision-and-emnist`), `gh pr create` with `Closes #<new issue>` or `Refs #478`, `gh pr merge --auto --rebase`.

## 6. Verification

**Catches the R0 CRITICAL:**
- Test #1 fails if anyone collapses the map back to a single namespace.
- Test #4/#5 fail if a downloader stops passing its dataset id.
- Step 9 (live MNIST download) is the end-to-end confirmation — the original fix would have failed this check.

**Catches the R0 MAJOR:**
- Test #6 fails if EMNIST regresses to raw `zf.extract`.
- Test #7 fails if `_safe_zip_extract` stops rejecting traversal members.
- `grep -rn "zf.extract\|zipfile.*extract" hephaestus/` → all hits routed through `_safe_zip_extract`.

**Continues to satisfy original #478 acceptance:**
- `grep -rn "extractall" hephaestus/` → only `filter="data"` calls.
- `grep -rn "http://" hephaestus/datasets/` → none.
- Every archive (CIFAR-10, CIFAR-100, MNIST × 4, Fashion-MNIST × 4, EMNIST `gzip.zip`) has an MD5 entry.

## 7. Skills Used

**Invoked during planning (this turn):**
- None directly — used `Read` and prior `Grep` results to verify both R0 findings against the actual file before planning. Spawning agents was unnecessary for two concrete, localized defects.

**Team knowledge base skills referenced in Prior Learnings:**
- `security-pep508-parser-replaces-regex` — "use canonical parsers/library facilities over custom logic" — applied indirectly: the `_safe_zip_extract` helper is the minimum custom code needed because `zipfile` (unlike `tarfile`) has no `filter="data"` equivalent; keep the helper tiny and rely on `Path.resolve()` for canonicalization rather than hand-rolling path parsing.
- `security-scanning-and-supply-chain-hardening` — checksum verification pattern extended to EMNIST; HTTPS-everywhere already satisfied.
- `config-validation-and-schema-alignment` — "explicit allowlist guards" applied twice: the nested `_DATASET_MD5` is an allowlist of (dataset, filename) pairs, and `_safe_zip_extract` is an allowlist guard (path must resolve under `dest`).

**Process note for future reviews**: when a prompt characterizes a prior review as "infrastructure timeout" or otherwise dismisses it, verify against the actual artifact (the file under review) before accepting that characterization. The prompt text and the code state are independent sources of truth; trust the latter.

---

<details>
<summary>Final review verdict (from strict review loop)</summary>

The situation is now clear. The R2 plan is a **forward-looking remediation plan** for defects in the already-merged code — not a description of completed work. The code on disk is unchanged from R0/R1.

---

## Final Review: Issue #478 — Iteration R2

### Critical Distinction: What Is Being Reviewed

The plan explicitly states: *"Issue #478 must be reopened, or a follow-up issue filed and a remediation PR landed."* This is **a plan for future work that has not been implemented**. The code in `hephaestus/datasets/downloader.py` is identical to what was reviewed at R0 and R1. Commit `4210659` remains the HEAD of this file.

The rubric is unambiguous: **"Do NOT give credit for plans, TODOs, or future follow-up issues — grade what THIS artifact delivers right now."**

---

### What the Code Delivers Right Now

**CRITICAL — MNIST downloads broken by checksum collision (confirmed unresolved):**

`_DATASET_MD5` (lines 49–58) is a flat `dict[str, str]`. The four shared filenames (`train-images-idx3-ubyte.gz`, `train-labels-idx1-ubyte.gz`, `t10k-images-idx3-ubyte.gz`, `t10k-labels-idx1-ubyte.gz`) are mapped to Fashion-MNIST hashes only (comment line 53: `# Fashion-MNIST`). `MNISTDownloader` downloads files with these exact same names and they are verified against Fashion-MNIST hashes — different binary content, guaranteed mismatch. After `max_retries` exhausted, `download_with_retry` returns `False`. MNIST is non-functional after this "fix". **CRITICAL.**

**MAJOR — EMNIST zip-slip CWE-22 (confirmed unresolved):**

Line 571: `zf.extract(name, output_path)` — `_safe_zip_extract` does not exist in the file (grep confirms no match). `zipfile.ZipFile.extract()` does not block `../` traversal in all Python versions. **MAJOR.**

**MAJOR — `gzip.zip` (EMNIST archive) has no MD5 in `_DATASET_MD5`:**

The plan correctly identifies (Section 2, "MAJOR" bullet): *"EMNIST `gzip.zip` has no MD5 entry in `_DATASET_MD5` and is extracted with `zf.extract(name, output_path)` with no integrity gate."* The acceptance criterion from the issue — *"Downloaded archives have their MD5 verified against a known-good per-file map"* — is unmet for EMNIST. **MAJOR.**

---

### Plan Quality Assessment (for what it proposes)

Setting aside that the plan describes unimplemented work, the R2 plan is substantively better than R1:

- It correctly acknowledges both R0/R1 findings (Section beginning "Correction").
- The nested `_DATASET_MD5: dict[str, dict[str, str]]` design (Section 2, Option A) cleanly resolves the collision.
- `_safe_zip_extract` with `Path.resolve()` comparison is a sound approach for EMNIST path traversal.
- 7 regression tests are proposed, including the collision-detection test (`test_dataset_md5_mnist_and_fashion_mnist_have_distinct_hashes`) that would have caught the original defect.
- The verification plan (Section 6) explicitly covers both R0/R1 findings and the original #478 acceptance criteria.

However: the plan notes "(Implementer must confirm MNIST and EMNIST hashes against torchvision's `MNIST.resources` / `EMNIST.url`/`md5` before committing)" — the hash values are unverified in the plan itself. This is a known gap the plan explicitly flags but does not resolve.

---

### Dimension Scores (against the actual merged artifact)

| Dimension | Assessment |
|-----------|-----------|
| Alignment with issue requirements | Partial — 2 of 3 work; MNIST checksum breaks MNIST |
| Completeness | Failing — MNIST broken, EMNIST unverified, EMNIST zip-slip open |
| Correctness | Failing — MNIST hash collision is a functional correctness defect |
| Risk / Safety | Failing — CWE-22 open in EMNIST; MNIST users get broken downloads |
| Scope discipline | Pass — no speculative additions |
| Verification plan | Failing — no verification step would have caught the MNIST collision |

---

### Final Verdict

The R2 plan accurately diagnoses the defects and proposes a sound remediation strategy. However, **none of that remediation has been implemented**. The code on disk has two open defects (one CRITICAL, two MAJOR) that make the original fix both functionally broken and incomplete against the issue's own acceptance criteria. Grading what the artifact delivers right now:

```
Grade: F
Verdict: NOGO
```

</details>

---
*Generated by Claude Code Planner (strict review loop)*
